ddl/ddl_api.go

```diff
@@ -72,10 +72,81 @@ const (
+	var opt *ast.CharsetOpt
+	if len(dbOps) != 0 {
+		opt = &ast.CharsetOpt{}
+		for _, val := range dbOps {
+			switch val.Tp {
+			case ast.DatabaseOptionCharset:
+				opt.Chs = val.Value
+			case ast.DatabaseOptionCollate:
+				opt.Col = val.Value
+			case ast.DatabaseOptionPlacementPrimaryRegion:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.PrimaryRegion = val.Value
+			case ast.DatabaseOptionPlacementRegions:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.Regions = val.Value
+			case ast.DatabaseOptionPlacementFollowerCount:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.Followers = val.UintValue
+			case ast.DatabaseOptionPlacementVoterCount:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.Voters = val.UintValue
+			case ast.DatabaseOptionPlacementLearnerCount:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.Learners = val.UintValue
+			case ast.DatabaseOptionPlacementSchedule:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.Schedule = val.Value
+			case ast.DatabaseOptionPlacementConstraints:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.Constraints = val.Value
+			case ast.DatabaseOptionPlacementLeaderConstraints:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.LeaderConstraints = val.Value
+			case ast.DatabaseOptionPlacementLearnerConstraints:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.LearnerConstraints = val.Value
+			case ast.DatabaseOptionPlacementFollowerConstraints:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.FollowerConstraints = val.Value
+			case ast.DatabaseOptionPlacementVoterConstraints:
+				if dbInfo.DirectPlacementOpts == nil {
+					dbInfo.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				dbInfo.DirectPlacementOpts.VoterConstraints = val.Value
+			case ast.DatabaseOptionPlacementPolicy:
+				dbInfo.PlacementPolicyRef = &model.PolicyRefInfo{
+					Name: model.NewCIStr(val.Value),
+				}
+			}
+		}
+	}
+	if opt != nil {
+		chs, coll, err := ResolveCharsetCollation(ast.CharsetOpt{Chs: opt.Chs, Col: opt.Col})
```

Comment:
I think some tests are still passing charset https://github.com/pingcap/tidb/blob/310fef18ed18b53d80c524a09277984532c7ff89/domain/domain_test.go#L136. You did not take care of all occurrences of the function.

And now `opt` is always non-nil even if no charset options are passed. That is a different logic than the original logic. @xiongjiwei , PTAL